File: 37-fenwick-tree/rust/src/main.rs
use serde::{Deserialize, Serialize};
use std::io::{self, Read};

#[derive(Deserialize)]
struct Input {
    n: usize,
    initial: Vec<i64>,
    queries: Vec<Vec<serde_json::Value>>,
}

fn solve(n: usize, initial: &[i64], queries: &[Vec<serde_json::Value>]) -> Vec<i64> {
    // TODO: Fenwick tree — handle ["sum", i] and ["update", i, delta] queries
    // Return results of sum queries in order
    let _ = (n, initial, queries);
    todo\!()
}

fn main() {
    let mut buf = String::new();
    io::stdin().read_to_string(&mut buf).unwrap();
    let inp: Input = serde_json::from_str(&buf).unwrap();
    let result = solve(inp.n, &inp.initial, &inp.queries);
    println\!("{}", result.iter().map(|v| v.to_string()).collect::<Vec<_>>().join(" "));
}

File: 37-fenwick-tree/python/main.py
import json
import sys


def solve(n, initial, queries):
    tree = [0] * (n + 1)

    def update(i, delta):
        while i <= n:
            tree[i] += delta
            i += i & (-i)

    def prefix_sum(i):
        s = 0
        while i > 0:
            s += tree[i]
            i -= i & (-i)
        return s

    for i, v in enumerate(initial, start=1):
        update(i, v)

    out = []
    for q in queries:
        if q[0] == "sum":
            out.append(prefix_sum(q[1]))
        else:
            update(q[1], q[2])
    return out


def main():
    obj = json.load(sys.stdin)
    print(*solve(obj["n"], obj["initial"], obj["queries"]))


if __name__ == "__main__":
    main()
